ui/src/tugboat/components/Saliency.tsx

```diff
@@ -0,0 +1,253 @@
+/**
+ * Renders a component that displays each word of the tokens with a different color based on the
+ * associated weight.
+ *
+ * There is also a slider used to decide a cutoff on what weights to not render.
+ */
+
+import React from 'react';
+import styled from 'styled-components';
+import colormap from 'colormap';
+import { Tooltip } from 'antd';
+
+interface DefaultProps {
+    colormapProps: ColorMapProps;
+}
+
+interface Props extends DefaultProps {
+    interpretData: [number[], number[]];
+    inputTokens: [string[], string[]];
+    inputHeaders: JSX.Element[];
+}
+
+interface State {
+    topK: { [id: string]: number };
+}
+
+interface ColorMapProps {
+    colormap: ColorScheme;
+    format: ColorMapFormat;
+    nshades: number;
+}
+
+type ColorMapFormat = 'hex' | 'rgbaString' | 'rba' | 'float';
+
+type ColorScheme =
+    | 'jet'
+    | 'hsv'
+    | 'hot'
+    | 'cool'
+    | 'spring'
+    | 'summer'
+    | 'autumn'
+    | 'winter'
+    | 'bone'
+    | 'copper'
+    | 'greys'
+    | 'YIGnBu'
+    | 'greens'
+    | 'YIOrRd'
+    | 'bluered'
+    | 'RdBu'
+    | 'picnic'
+    | 'rainbow'
+    | 'portland'
+    | 'blackbody'
+    | 'earth'
+    | 'electric'
+    | 'viridis'
+    | 'inferno'
+    | 'magma'
+    | 'plasma'
+    | 'warm'
+    | 'rainbow-soft'
+    | 'bathymetry'
+    | 'cdom'
+    | 'chlorophyll'
+    | 'density'
+    | 'freesurface-blue'
+    | 'freesurface-red'
+    | 'oxygen'
+    | 'par'
+    | 'phase'
+    | 'salinity'
+    | 'temperature'
+    | 'turbidity'
+    | 'velocity-blue'
+    | 'velocity-green'
+    | 'cubehelix';
+
+interface TokensWithWeight {
+    token: string;
+    weight: number;
+}
+
+export class Saliency extends React.Component<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            topK: { all: 3 }, // 3 words are highlighted by default
+        };
+    }
+
+    static defaultProps: DefaultProps = {
+        colormapProps: {
+            colormap: 'copper',
+            format: 'hex',
+            nshades: 20,
+        },
+    };
+
+    getTokenWeightPairs(grads: number[], tokens: string[]) {
+        return tokens.map((token, idx: number) => {
+            const weight = grads[idx];
+            // We do 1 - weight because the colormap is inverted
+            return { token, weight: 1 - weight };
+        });
+    }
+
+    colorize(tokensWithWeights: TokensWithWeight[], topKIdx: Set<number>) {
+        const { colormapProps } = this.props;
+        // colormap package takes minimum of 6 shades
+        colormapProps.nshades = Math.min(Math.max(colormapProps.nshades, 6), 72);
+        const colors = colormap(colormapProps);
```

Comment:
```suggestion
        const nshades = Math.min(Math.max(colormapProps.nshades, 6), 72);
        const colors = colormap({ ...colormapProps, nshades });
```

This avoids mutating the property (which is a reference, so anyone with a handle to it will receive any mutations).